### DISCUSSION TITLE
null 리턴은 왜 나쁠까?

### DISCUSSION THREAD

**Discussion by seunghye218** [2024-woowacourse-study/level-interview#75] [Clean Code]
.

**Comment by jimi567** [comment]
null 값을 리턴하게 될 경우 장단점을 다음과 같습니다.

장점
기본적으로 null을 리턴하는 것은 가장 간단한 방법입니다. 특히, 옵셔널 상황에서 반환할 값이 없을 때 사용하기 편합니다. 해당 메서드가 논리적으로 프로덕트 코드에서 호출될 수 없을 경우, 이를 의도적으로 나타내기도 편합니다.(널 객체 패턴)
단점
NPE가 발생할 가능성이 높아집니다. NPE의 경우 발생 시점을 추적하기 어려워 디버깅이나 유지 보수하기 어려운 코드가 될 수 있습니다.
또,  호출자의 null 체크가 확산될 가능성이 높기 때문에 코드의 가독성이나 복잡성을 증가 시키고, 코드 중복을 야기할 수 있습니다.

**Comment by nayonsoso** [comment]
null을 리턴한다면, Null Pointer Exception(NPE)가 발생할 수 있습니다.

NPE는 코드의 어느 부분에서 문제가 생겼는지 추척하기 어렵습니다.
코드를 작성하는 개발자가 NPE를 의도한 것이라고 해도,
동료 개발자들은 NPE가 의도적인 것인지, 어떤 의도를 담고 있는지 이해하기 위한 리소스가 필요합니다.

따라서 null을 리턴하기보다 다음의 대안을 사용하는 것이 좋다고 생각합니다.
1. '빈 객체'라는 의미를 전달하는 객체를 반환
2. Optional.empty를 사용
3. 예외를 발생시키기

**Comment by GIVEN53** [comment]
NPE
트래킹, 유지보수 어려움

대안
null 객체
Optional
예외 던짐
